pages/docs/couple-your-code/couple-your-code-waveform.md

```diff
@@ -0,0 +1,167 @@
+
+precice::SolverInterface precice("FluidSolver","precice-config.xml",rank,size); // constructor
+
+int dim = precice.getDimensions();
+int meshID = precice.getMeshID("FluidMesh");
+int vertexSize; // number of vertices at wet surface 
+// determine vertexSize
+double* coords = new double[vertexSize*dim]; // coords of coupling vertices 
+// determine coordinates
+int* vertexIDs = new int[vertexSize];
+precice.setMeshVertices(meshID, vertexSize, coords, vertexIDs); 
+delete[] coords;
+
+int displID = precice.getDataID("Displacements", meshID); 
+int forceID = precice.getDataID("Forces", meshID); 
+double* forces = new double[vertexSize*dim];
+double* displacements = new double[vertexSize*dim];
+
+double dt;         // solver timestep size
+double precice_dt; // maximum precice timestep size
+
+precice_dt = precice.initialize();
+while (not simulationDone()){ // time loop
+  dt = beginTimeStep(); // e.g. compute adaptive dt 
+  dt = min(precice_dt, dt);
+  // sampling in the middle of the time step
+  precice.readBlockVectorData(displID, vertexSize, vertexIDs, 0.5 * dt, displacements);
+  setDisplacements(displacements); // displacement at the middle of the time step
+  solveTimeStep(dt); // might be using midpoint rule for time-stepping
+  computeForces(forces);
+  precice.writeBlockVectorData(forceID, vertexSize, vertexIDs, forces);
+  precice_dt = precice.advance(dt);
+  endTimeStep(); // e.g. update variables, increment time
+}
+precice.finalize();
+delete[] vertexIDs, forces, displacements;
+turnOffSolver();
+```
+
+### Use waveform iteration with subcycling
+
+Another example shows how to use subcycling with the waveform iteration API. This is an extension of the example from ["Step 5 - Non-matching time step sizes"](couple-your-code-timestep-sizes.html). The changes are similar to the example above, but additionally the behavior of `isReadDataAvailable` slightly changes, because every time step now provides updated read data:
+
+```cpp
+while (not simulationDone()){ // time loop
+  dt = beginTimeStep(); // e.g. compute adaptive dt 
+  dt = min(precice_dt, dt);
+  if (precice.isReadDataAvailable()){ // always returns true with new API
+    // sampling in the middle of the time step
+    precice.readBlockVectorData(displID, vertexSize, vertexIDs, 0.5 * dt, displacements);
+    setDisplacements(displacements); // displacement at the middle of the time step
+  }
+  solveTimeStep(dt); // might be using midpoint rule for time-stepping
+  if (precice.isWriteDataRequired(dt)){ // only true at the end of the time window
+    computeForces(forces);
+    precice.writeBlockVectorData(forceID, vertexSize, vertexIDs, forces);
+  }
+  precice_dt = precice.advance(dt);
+  endTimeStep(); // e.g. update variables, increment time
+}
+```
+
+### Note on the use of `initializeData`
+
+{% include important.html content="`writeBlockVectorData` *after* `initializeData` writes data corresponding to the *end* of the window." %}
```

Comment:
This, I don't get. Seems also like a misuse of the "important" box?

Reply:
My idea here was to tell the reader that `initializeData` is automatically needed to set data at the beginning of the simulation, but I think this is clear and not needed, after our revision of how `initializeData` and waveform relaxation should work.

Reply:
Revised this in 8ba2573eab4a20317a4cc5c247d21438d9f8f284.